# Sprint 3 Tasca M3 T02
## Exercici 1
**Descarrega el data set Airlines Delay: Airline on-time statistics and delay causes i carrega’l a un Pandas Dataframe. Explora les dades que conté, explica breument quines variables hi ha i queda’t únicament amb les columnes que consideris rellevants. Justifica la teva elecció.**

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('data/DelayedFlights.csv', encoding='utf-8')
pd.set_option('display.max_columns', None)
pd.set_option('display.expand_frame_repr', False)

In [3]:
df.head()

Unnamed: 0  Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  CRSArrTime UniqueCarrier  FlightNum TailNum  ActualElapsedTime  CRSElapsedTime  AirTime  ArrDelay  DepDelay Origin Dest  Distance  TaxiIn  TaxiOut  Cancelled CancellationCode  Diverted  CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  LateAircraftDelay
0           0  2008      1           3          4   2003.0        1955   2211.0        2225            WN        335  N712SW              128.0           150.0    116.0     -14.0       8.0    IAD  TPA       810     4.0      8.0          0                N         0           NaN           NaN       NaN            NaN                NaN
1           1  2008      1           3          4    754.0         735   1002.0        1000            WN       3231  N772SW              128.0           145.0    113.0       2.0      19.0    IAD  TPA       810     5.0     10.0          0                N         0           NaN           NaN       NaN            NaN                NaN
2           2  2008      1           3          4    628.0         620    804.0         750            WN        448  N428WN               96.0            90.0     76.0      14.0       8.0    IND  BWI       515     3.0     17.0          0                N         0           NaN           NaN       NaN            NaN                NaN
3           4  2008      1           3          4   1829.0        1755   1959.0        1925            WN       3920  N464WN               90.0            90.0     77.0      34.0      34.0    IND  BWI       515     3.0     10.0          0                N         0           2.0           0.0       0.0            0.0               32.0
4           5  2008      1           3          4   1940.0        1915   2121.0        2110            WN        378  N726SW              101.0           115.0     87.0      11.0      25.0    IND  JAX       688     4.0     10.0          0                N         0           NaN           NaN       NaN            NaN                NaN

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1936758 entries, 0 to 1936757
Data columns (total 30 columns):
 #   Column             Dtype  
---  ------             -----  
 0   Unnamed: 0         int64  
 1   Year               int64  
 2   Month              int64  
 3   DayofMonth         int64  
 4   DayOfWeek          int64  
 5   DepTime            float64
 6   CRSDepTime         int64  
 7   ArrTime            float64
 8   CRSArrTime         int64  
 9   UniqueCarrier      object 
 10  FlightNum          int64  
 11  TailNum            object 
 12  ActualElapsedTime  float64
 13  CRSElapsedTime     float64
 14  AirTime            float64
 15  ArrDelay           float64
 16  DepDelay           float64
 17  Origin             object 
 18  Dest               object 
 19  Distance           int64  
 20  TaxiIn             float64
 21  TaxiOut            float64
 22  Cancelled          int64  
 23  CancellationCode   object 
 24  Diverted           int64  
 25  CarrierDelay      

**Unnamed: 0**: Una columna con un índice numérico de los registros.

**Year**: El año de la fecha del vuelo.

**Month**: El mes de la fecha del vuelo.

**DayofMonth**: El día del mes de la fecha del vuelo.

**DayOfWeek**: El día de la semana de la fecha del vuelo, donde 1 es lunes y 7 es domingo.

**DepTime**: La hora de salida real del vuelo.

**CRSDepTime**: La hora de salida programada del vuelo.

**ArrTime**: La hora de llegada real del vuelo.

**CRSArrTime**: La hora de llegada programada del vuelo.

**UniqueCarrier**: El código de identificación único de la aerolínea.

**FlightNum**: El número de vuelo.

**TailNum**: El número de cola de la aeronave.

**ActualElapsedTime**: Duración real del vuelo en minutos.

**CRSElapsedTime**: Duración del vuelo programado en minutos.

**AirTime**: El tiempo de vuelo en el aire en minutos.

**ArrDelay**: El retraso en la llegada en minutos.

**DepDelay**: El retraso en la salida en minutos.

**Origin**: El aeropuerto de origen del vuelo.

**Dest**: El aeropuerto de destino del vuelo.

**Distance**: La distancia del vuelo en millas.

**TaxiIn**: El tiempo de rodaje en la llegada en minutos.

**TaxiOut**: El tiempo de rodaje en la salida en minutos.

**Cancelled**: Indica si el vuelo fue cancelado (1) o no (0).

**CancellationCode**: El código de cancelación si el vuelo fue cancelado.

**Diverted**: Indica si el vuelo fue desviado a otro aeropuerto (1) o no (0).

**CarrierDelay**: El tiempo de retraso debido a la aerolínea en minutos.

**WeatherDelay**: El tiempo de retraso debido al clima en minutos.

**NASDelay**: El tiempo de retraso debido al Sistema Nacional de Aviación en minutos.

**SecurityDelay**: El tiempo de retraso debido a problemas de seguridad en minutos.

**LateAircraftDelay**: El tiempo de retraso debido a problemas con el avión en minutos.

Inicialmente, considero todas las columnas relevantes. Probablemente no las mantendría todas en un solo dataframe, pero en principio encuentro que todas son interesantes para su uso, ya que podemos obtener datos importantes de cada una. La única excepción a primera vista sería la columna 'Unnamed: 0', ya que un índice no será útil para el análisis.
Tengo la duda de si la columna 'Year' tiene valores de más de un año o si todos son del mismo año. Procedamos a mirarlo.

In [5]:
df.Year.value_counts()

2008    1936758
Name: Year, dtype: int64


Dado que todos los registros son de 2008, podemos proceder a eliminar la columna, ya que no aporta información relevante.
Antes de continuar, verifiquemos la presencia de valores nulos para identificar columnas repletas de registros corruptos o inexistentes. De esta manera, poderlas eliminar del dataframe y evitar llevar información no deseada.

In [6]:
valores_nulos = df.isnull().sum()

print(valores_nulos)

Unnamed: 0                0
Year                      0
Month                     0
DayofMonth                0
DayOfWeek                 0
DepTime                   0
CRSDepTime                0
ArrTime                7110
CRSArrTime                0
UniqueCarrier             0
FlightNum                 0
TailNum                   5
ActualElapsedTime      8387
CRSElapsedTime          198
AirTime                8387
ArrDelay               8387
DepDelay                  0
Origin                    0
Dest                      0
Distance                  0
TaxiIn                 7110
TaxiOut                 455
Cancelled                 0
CancellationCode          0
Diverted                  0
CarrierDelay         689270
WeatherDelay         689270
NASDelay             689270
SecurityDelay        689270
LateAircraftDelay    689270
dtype: int64


Tras analizar los resultados, realizaremos un breve estudio de las últimas 5 columnas, ya que aproximadamente uno de cada tres registros contiene un valor nulo. Es posible que, al estar relacionadas entre sí, durante la toma de registros se haya registrado el retraso pertinente y se hayan dejado vacías las demás columnas. Esto se podría solucionar fácilmente convirtiendo los valores 'NaN' en '0'.

In [7]:
columnas_busqueda = ['CarrierDelay', 'WeatherDelay', 'NASDelay', 'SecurityDelay','SecurityDelay', 'LateAircraftDelay' ]

registros_filtrados = df[df[columnas_busqueda].isnull().any(axis=1)]

registros_filtrados.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 689270 entries, 0 to 1936757
Data columns (total 30 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   Unnamed: 0         689270 non-null  int64  
 1   Year               689270 non-null  int64  
 2   Month              689270 non-null  int64  
 3   DayofMonth         689270 non-null  int64  
 4   DayOfWeek          689270 non-null  int64  
 5   DepTime            689270 non-null  float64
 6   CRSDepTime         689270 non-null  int64  
 7   ArrTime            682160 non-null  float64
 8   CRSArrTime         689270 non-null  int64  
 9   UniqueCarrier      689270 non-null  object 
 10  FlightNum          689270 non-null  int64  
 11  TailNum            689267 non-null  object 
 12  ActualElapsedTime  680883 non-null  float64
 13  CRSElapsedTime     689072 non-null  float64
 14  AirTime            680883 non-null  float64
 15  ArrDelay           680883 non-null  float64
 16  D

Después de realizar esta comprobación, observamos que los 689,270 registros no contienen ningún valor en esas 5 columnas. La única explicación que se me ocurre es que estos registros correspondan a vuelos sin ningún retraso, vuelos cancelados o vuelos redirigidos, lo cual podría justificar por qué esas columnas no están completadas.

In [8]:
registros_con_arrdelay_cero = df[df['ArrDelay'] <= 0]

registros_con_arrdelay_cero.shape[0]

204956

In [9]:
registros_filtrados.Cancelled.value_counts()

0    688637
1       633
Name: Cancelled, dtype: int64

In [10]:
registros_filtrados.Diverted.value_counts()

0    681516
1      7754
Name: Diverted, dtype: int64

Después de analizar los datos, se observa que la suma de los casos posibles alcanza poco más de 213,000 de los casi 690,000 registros disponibles. Esto implica que todavía hay alrededor de 500,000 registros restantes de los casi 2,000,000 registros originales. Aproximadamente 1 de cada 4 registros no tiene completadas estas columnas sin una razón aparente.
La única idea mínimamente válida para completar estos datos habría sido establecer en cero los registros de vuelos sin retraso, cancelados o desviados, y en el resto, dividir por 5 el valor de 'ArrDelay' y asignar el mismo valor a todos, suponiendo que la responsabilidad se reparte por igual. Sin embargo, a pesar de esta posible solución, prefiero eliminar estas columnas para los siguientes ejercicios.  Con esta información en mente, he decidido eliminar estas columnas del dataframe para evitar trabajar con datos incompletos que podrían distorsionar los cálculos de medias y análisis subsiguientes. En caso de necesitar trabajar con ellas en el futuro, se puede crear un nuevo dataframe solo con los registros válidos.
En base a este estudio, se observa que la suma de vuelos cancelados y desviados es igual a 8387, que es el número de valores nulos presentes en las variables 'AirTime', 'ArrDelay' y 'ActualElapsedTime'. Es comprensible que no se disponga de datos para estos vuelos, lo cual plantea la opción de separar estos datos en otro dataframe y eliminarlos del actual para realizar un análisis estadístico más preciso sobre los retrasos en los vuelos. Para lograr esto, simplemente debemos confirmar que esos 8387 valores nulos corresponden exactamente a los vuelos cancelados y desviados, y no son producto de la casualidad.

In [11]:
columna_busqueda = ['AirTime', 'ArrDelay', 'ActualElapsedTime']

registro_filtrado = df[df[columna_busqueda].isnull().any(axis=1)]

In [12]:
registro_filtrado.Cancelled.value_counts()

0    7754
1     633
Name: Cancelled, dtype: int64

In [13]:
registro_filtrado.Diverted.value_counts()

1    7754
0     633
Name: Diverted, dtype: int64

Efectivamente, todos los valores nulos en esas columnas corresponden a los vuelos cancelados y retrasados presentes en el dataframe. Por lo tanto, procederemos a eliminarlos de nuestros registros. Sin embargo, antes de hacerlo, por curiosidad y para obtener una medida estadística, podemos calcular el porcentaje que representan estos vuelos con respecto al total.

In [14]:
total = len(df)

cancelled = len(df[df['Cancelled'] == 1]) 
cancelled_percent = cancelled/total * 100

diverted = len(df[df['Diverted'] == 1])
diverted_percent = diverted/total * 100

print("Cancelled: {:.2f}%".format(cancelled_percent))
print("Diverted: {:.2f}%".format(diverted_percent))

Cancelled: 0.03%
Diverted: 0.40%


Una vez vistas las estadisticas procedemos a crear un dataframe nuevo sin estos registros.

In [15]:
df_limpio = df.dropna(subset=columna_busqueda)

valores_nulos = df_limpio.isnull().sum()
print(valores_nulos)

Unnamed: 0                0
Year                      0
Month                     0
DayofMonth                0
DayOfWeek                 0
DepTime                   0
CRSDepTime                0
ArrTime                   0
CRSArrTime                0
UniqueCarrier             0
FlightNum                 0
TailNum                   3
ActualElapsedTime         0
CRSElapsedTime            0
AirTime                   0
ArrDelay                  0
DepDelay                  0
Origin                    0
Dest                      0
Distance                  0
TaxiIn                    0
TaxiOut                   0
Cancelled                 0
CancellationCode          0
Diverted                  0
CarrierDelay         680883
WeatherDelay         680883
NASDelay             680883
SecurityDelay        680883
LateAircraftDelay    680883
dtype: int64


Después de llegar a este punto, solo nos quedan 3 valores nulos en la columna 'NailNum''. La verdad es que pensé que podía corresponder a un vuelo regular de línea y poner el número de cola del avión más utilizado, pero no es el caso :(
A continuación, procederemos a eliminar las columnas mencionadas anteriormente: 'Cancelled', 'CancellationCode' y 'Diverted', ya que no las necesitaremos. Posteriormente, reduciremos aleatoriamente el conjunto de datos de casi dos millones de registros a 200,000, con la esperanza de que los valores nulos en la columna 'NailNum' queden excluidos :)

In [16]:
columnas_a_eliminar = ['Unnamed: 0', 'Year', 'CarrierDelay', 'WeatherDelay', 'NASDelay', 'SecurityDelay', 'LateAircraftDelay',  'Cancelled', 'CancellationCode', 'Diverted']
df_limpio = df_limpio.drop(columns=columnas_a_eliminar)
df_limpio.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1928371 entries, 0 to 1936757
Data columns (total 20 columns):
 #   Column             Dtype  
---  ------             -----  
 0   Month              int64  
 1   DayofMonth         int64  
 2   DayOfWeek          int64  
 3   DepTime            float64
 4   CRSDepTime         int64  
 5   ArrTime            float64
 6   CRSArrTime         int64  
 7   UniqueCarrier      object 
 8   FlightNum          int64  
 9   TailNum            object 
 10  ActualElapsedTime  float64
 11  CRSElapsedTime     float64
 12  AirTime            float64
 13  ArrDelay           float64
 14  DepDelay           float64
 15  Origin             object 
 16  Dest               object 
 17  Distance           int64  
 18  TaxiIn             float64
 19  TaxiOut            float64
dtypes: float64(9), int64(7), object(4)
memory usage: 309.0+ MB


Ahora es tiempo de mirar si existen duplicados y eliminarlos si los hubiera.


In [17]:
duplicados = df_limpio.duplicated()
print(duplicados.sum())


2


In [18]:
df_limpio = df_limpio.drop_duplicates()

A continuación reduciremos el dataset a 200000 registros de manera aleatoria.

In [19]:
df_reducido = df_limpio.sample(n=200000, random_state=23)

df_reducido

Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  CRSArrTime UniqueCarrier  FlightNum TailNum  ActualElapsedTime  CRSElapsedTime  AirTime  ArrDelay  DepDelay Origin Dest  Distance  TaxiIn  TaxiOut
195608       2          12          2    839.0         820    921.0         910            WN       1952  N390SW              102.0           110.0     88.0      11.0      19.0    DAL  ABQ       580     4.0     10.0
1558072     10          30          4   1615.0        1603   1754.0        1717            OO       5428  N234SW               99.0            74.0     88.0      37.0      12.0    ACV  SFO       250     5.0      6.0
1581600     10           8          3   1340.0        1157   1602.0        1423            EV       5188  N633AS               82.0            86.0     65.0      99.0     103.0    PFN  ATL       247     7.0     10.0
1491875      9           8          1   1041.0        1035   1151.0        1145            MQ       3843  N641AE               70.0            70.0     53.0       6.0       6.0    DFW  LIT       304     4.0     13.0
1129086      7          31          4   1113.0        1025   1205.0        1114            YV       2891  N909FJ               52.0            49.0     24.0      51.0      48.0    TUS  PHX       110    13.0     15.0
...        ...         ...        ...      ...         ...      ...         ...           ...        ...     ...                ...             ...      ...       ...       ...    ...  ...       ...     ...      ...
1678011     11           2          7   2232.0        2130     49.0        2350            WN       1082  N221WN               77.0            80.0     58.0      59.0      62.0    LAS  SLC       368     5.0     14.0
1257174      7           3          4   1801.0        1755   1901.0        1857            CO       1779  N16732               60.0            62.0     40.0       4.0       6.0    IAH  SAT       191     3.0     17.0
1661715     11          24          1   1130.0         854   1313.0        1039            OO       5769  N234SW              103.0           105.0     85.0     154.0     156.0    LMT  SFO       315     3.0     15.0
1003913      6          23          1   1448.0        1442   1641.0        1645            FL        261  N958AT              113.0           123.0     97.0      -4.0       6.0    FNT  ATL       644     7.0      9.0
553695       3          17          1     19.0        2150    236.0          30            B6         58  N583JB              137.0           160.0    120.0     126.0     149.0    MCO  JFK       944     6.0     11.0

[200000 rows x 20 columns]

# Exercici 2
**Fes un informe complet del dataset:**

- **Resumeix estadísticament el dataset i les columnes d’interès.**

In [20]:
# resumen estadístico de las variables numéricas

df_reducido.describe().T

count         mean          std   min     25%     50%     75%     max
Month              200000.0     6.113320     3.483689   1.0     3.0     6.0     9.0    12.0
DayofMonth         200000.0    15.767080     8.772620   1.0     8.0    16.0    23.0    31.0
DayOfWeek          200000.0     3.987190     1.995541   1.0     2.0     4.0     6.0     7.0
DepTime            200000.0  1517.837595   450.742260   1.0  1202.0  1545.0  1900.0  2400.0
CRSDepTime         200000.0  1467.473640   424.612892   1.0  1135.0  1510.0  1815.0  2359.0
ArrTime            200000.0  1607.728820   549.096988   1.0  1314.0  1714.0  2029.0  2400.0
CRSArrTime         200000.0  1633.624325   464.524840   0.0  1325.0  1705.0  2013.0  2359.0
FlightNum          200000.0  2185.611025  1943.550539   1.0   612.0  1546.0  3423.0  7829.0
ActualElapsedTime  200000.0   133.336515    72.126327  15.0    80.0   116.0   165.0   711.0
CRSElapsedTime     200000.0   134.244685    71.261507  15.0    81.0   117.0   165.0   660.0
AirTime            200000.0   108.294965    68.660097   2.0    58.0    90.0   137.0   655.0
ArrDelay           200000.0    42.123795    57.017990 -62.0     9.0    24.0    56.0  2461.0
DepDelay           200000.0    43.031955    53.528544   6.0    12.0    24.0    53.0  2467.0
Distance           200000.0   765.201745   573.786134  30.0   338.0   606.0   998.0  4962.0
TaxiIn             200000.0     6.801115     5.333447   0.0     4.0     5.0     8.0   225.0
TaxiOut            200000.0    18.240435    14.399869   0.0    10.0    14.0    21.0   386.0

- **Fes una anàlisi estadístic del que consideris rellevant.**

Al realizar el estudio estadístico, lo primero que busqué fue la presencia de registros absurdos, como un día de la semana superior a 7, una distancia negativa o exageradamente grande, una hora menor a 0 o mayor a 2400, entre otros. Sin embargo, parece que no existen valores corruptos en los datos. A continuación, haré un breve resumen por cada bloque de información.

En las columnas correspondientes al mes, día del mes y día de la semana, encontramos los valores esperados. En cuanto al día de la semana, la media es aproximadamente 4 y la desviación estándar indica que los valores están bastante dispersos. Con respecto al día del mes, ocurre algo similar, con una media ligeramente superior a 15.5 y una distribución dispersa. Es curioso notar que en la columna del mes del año, la media es casi 6, aunque debería estar más cercana a 6.5. Esto podría sugerir que hay ligeramente más vuelos en la primera mitad del año que en la segunda.

En relación a las columnas DepTime, CRSDepTime, ArrTime y CRSArrTime, simplemente podemos comentar que el estudio estadístico nos proporciona una idea aproximada. Sin embargo, la forma en que se tratan las horas en estas columnas muestra valores incoherentes, como el caso de CRSDepTime que la media es 1467. Sería interesante crear nuevas columnas que representaran las horas en formato decimal, lo cual tendrá un mayor valor estadístico. Vemos igualmente que las medias son las centrales del día teniendo en cuenta que a ciertas horas de la noche no se programan vuelos.

La columna FlightNum no tiene relevancia estadística en sí misma. Quizás podríamos calcular la moda o simplemente identificar los números de vuelo más repetidos.

En cuanto a ActualElapsedTime, CRSElapsedTime y AirTime, observamos que la duración promedio de los vuelos es de aproximadamente 2 horas, pero existe una gran variabilidad, desde 15 minutos hasta 11 horas. Lo que destaca es la similitud entre el tiempo estimado y el real, lo que indica que una vez que el avión tiene la orden de salida, los tiempos están meticulosamente medidos y planificados para minimizar el tiempo en el aire.

Los retrasos, que son la parte más interesante del dataframe, muestran medias similares entre los retrasos en las salidas y en las llegadas. Como mencionamos anteriormente, no es posible recortar significativamente los tiempos en el aire ni prolongarlos en exceso. Por lo tanto, un retraso en la salida probablemente conllevará un retraso en la llegada. Sin embargo, los valores de los retrasos en las llegadas son más variables, incluso encontramos valores negativos que indican que los vuelos han llegado antes de la hora programada. Y como es lógico, ningún vuelo despega antes de su hora.

La distancia es un valor muy variable, que va desde los 30 kilómetros hasta casi 5000 kilómetros en el vuelo más largo. Sin embargo, si observamos los percentiles, podemos ver que el 50% de los vuelos se encuentran en el rango de 300 a 1000 kilómetros.

Para finalizar, en cuanto a los tiempos de rodaje, llama la atención a primera vista que los tiempos en las llegadas son tres veces más largos que los de salida. Esto se explica por la prioridad de los despegues en los aeropuertos, es decir, dejen salir antes de entrar :-) Añadir simplemente que el 75% de los vuelos tienen horquillas de tiempo de rodaje muy ajustadas, mientras que existe una parte correspondientes a retrasos que pueden superar las 6 horas.


- **Troba quantes dades faltants hi ha per columna.**

In [21]:
valores_nulos = df_reducido.isnull().sum()

print(valores_nulos)

Month                0
DayofMonth           0
DayOfWeek            0
DepTime              0
CRSDepTime           0
ArrTime              0
CRSArrTime           0
UniqueCarrier        0
FlightNum            0
TailNum              0
ActualElapsedTime    0
CRSElapsedTime       0
AirTime              0
ArrDelay             0
DepDelay             0
Origin               0
Dest                 0
Distance             0
TaxiIn               0
TaxiOut              0
dtype: int64


Nos damos cuenta de que la suerte estuvo de nuestro lado y los posibles valores nulos quedaron fuera. 
En caso de que hubiera quedado alguno, habríamos procedido a eliminarlo directamente o, en su defecto, reemplazarlo por un '0' para no perder los datos.

- **Crea columnes noves (velocitat mitjana del vol, si ha arribat tard o no...).**

In [22]:
df_reducido['AvgSpeed'] = df_reducido['Distance'] / (df_reducido['AirTime']/60)
df_reducido['AvgSpeed'].describe()

count    200000.000000
mean        396.869368
std          77.954784
min          36.000000
25%         351.840000
50%         404.062500
75%         448.106312
max        6690.000000
Name: AvgSpeed, dtype: float64

In [23]:
Velocidad_max = df_reducido[df_reducido['AvgSpeed'] > 1000]
Velocidad_max

Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  CRSArrTime UniqueCarrier  FlightNum TailNum  ActualElapsedTime  CRSElapsedTime  AirTime  ArrDelay  DepDelay Origin Dest  Distance  TaxiIn  TaxiOut     AvgSpeed
1401975      8          29          5   1559.0        1505   1658.0        1600            AA       2378  N4WWAA               59.0            55.0      3.0      58.0      54.0    DFW  AUS       190     7.0     49.0  3800.000000
53537        1           5          6   1842.0        1715   1949.0        1824            OO       3694  N581SW               67.0            69.0      2.0      85.0      87.0    SLC  SUN       223     2.0     63.0  6690.000000
101588       1           6          7   2157.0        2142   2233.0        2220            EV       4277  N883AS               96.0            98.0     17.0      13.0      15.0    ATL  BTR       449    58.0     21.0  1584.705882
106705       1          31          4   1519.0        1459   1823.0        1753            EV       4769  N931EV              124.0           114.0     26.0      30.0      20.0    XNA  ATL       589    13.0     85.0  1359.230769
915054       6          27          5   1444.0        1415   1525.0        1505            WN       3686  N689SW               41.0            50.0      5.0      20.0      29.0    DAL  AUS       189    25.0     11.0  2268.000000
1383147      8          28          4   1802.0        1755   2000.0        1913            9E       2021  91479E              118.0            78.0      6.0      47.0       7.0    CLT  ATL       227    50.0     62.0  2270.000000
774620       5          14          3   2139.0        2005   2235.0        2219            OH       5458  N957CA               56.0           134.0     28.0      16.0      94.0    ATL  PHL       665    10.0     18.0  1425.000000
438251       3           9          7   1141.0         843   1318.0         929            OO       5406  N292SW               97.0            46.0      3.0     229.0     178.0    SAN  LAX       109    87.0      7.0  2180.000000
1308503      8          29          5    820.0         755   1010.0        1035            YV       2668  N914FJ               50.0           100.0     16.0     -25.0      25.0    MEM  CLT       512    10.0     24.0  1920.000000

A partir de la velocidad media, se puede deducir que algunos valores de tiempo en aire no son correctos, ya que resulta absurdo que un avión haya estado en el aire solo un minuto y haya recorrido 274 millas. En cuanto a los valores bajos, esto podría atribuirse a vuelos de corta distancia que estaban volando cerca del destino, esperando permiso para el aterrizaje.

In [24]:
df_reducido['Delayed'] = df_reducido['ArrDelay'] > 0
porcentaje_retrasos = (df_reducido['Delayed'].sum() / len(df_reducido)) * 100
print("Delayed: {:.2f}%".format(porcentaje_retrasos))

Delayed: 89.36%


Observamos que 11% llegan al destino antes de lo previsto y que el resto acumula algún tipo de retraso.
A continuación vamos a crear una nueva columna que sea 'DayofYear' la cual representará el número del dia en el presente año y así por ejemplo poder graficar comodamente la evolución a lo largo del año del número de vuelos diarios.

In [25]:
def calcular_day_of_year(row):
    month = row['Month']
    day = row['DayofMonth']
    
    days_in_month = [31, 29, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]
    day_of_year = sum(days_in_month[:month-1]) + day
    
    return day_of_year

df_reducido['DayofYear'] = df_reducido.apply(calcular_day_of_year, axis=1)
df_reducido.DayofYear.value_counts()

2      1089
357    1085
336    1073
354    1035
356    1024
       ... 
267     195
264     193
278     183
244     179
332     158
Name: DayofYear, Length: 366, dtype: int64

- **Fes una taula de les aerolínies amb més endarreriments acumulats.**

In [26]:
tabla_retrasos = df_reducido.groupby('UniqueCarrier')['Delayed'].sum().reset_index()
tabla_retrasos = tabla_retrasos.sort_values('Delayed', ascending=False)
print(tabla_retrasos.to_string(index=False))

UniqueCarrier  Delayed
           WN    33828
           AA    17972
           MQ    13528
           UA    12721
           OO    12700
           DL    10512
           XE     9880
           CO     8673
           US     8445
           EV     7785
           NW     7446
           FL     6653
           YV     6556
           OH     5122
           B6     5009
           9E     4821
           AS     3540
           F9     2690
           HA      774
           AQ       58


In [27]:
tabla_total_vuelos = df_reducido.groupby('UniqueCarrier').size().reset_index(name='TotalFlights')
tabla_retrasos = df_reducido.groupby('UniqueCarrier')['Delayed'].sum().reset_index(name='DelayedFlights')
tabla_retrasos['DelayedPercentage'] = (tabla_retrasos['DelayedFlights'] / tabla_total_vuelos['TotalFlights']) * 100
tabla_retrasos = tabla_retrasos.sort_values('DelayedPercentage', ascending=False)
print(tabla_retrasos.to_string(index=False))

UniqueCarrier  DelayedFlights  DelayedPercentage
           HA             774          96.149068
           YV            6556          95.042041
           OH            5122          93.929947
           OO           12700          92.531876
           MQ           13528          92.524451
           EV            7785          92.032155
           NW            7446          91.575452
           XE            9880          91.346154
           F9            2690          91.186441
           FL            6653          90.950103
           9E            4821          90.586246
           AA           17972          90.420608
           DL           10512          88.432742
           AS            3540          88.389513
           UA           12721          87.821885
           B6            5009          87.447626
           WN           33828          86.390684
           AQ              58          85.294118
           US            8445          84.542997
           CO       

Tenía que limpiar el nombre de SouthWest Airlines ('WN') puesto que acumula mas retrasos en terminos absolutos.Sin embargo si miramos los porcentajes respecto al total de vuelos, está entre las 3 aerolineas que acumulan menos retrasos.

- **Quins són els vols més llargs? I els més endarrerits?**

En esta sección, tomaremos en cuenta la duración de los vuelos para identificar cuáles son los más largos.

In [28]:
tabla_duracion = df_reducido.groupby('FlightNum')['CRSElapsedTime'].max().reset_index()
tabla_duracion = tabla_duracion.sort_values('CRSElapsedTime', ascending=False).head(5)

print(tabla_duracion.to_string(index=False))


 FlightNum  CRSElapsedTime
        15           660.0
      1561           604.0
      1273           604.0
       851           601.0
        14           565.0


In [29]:
tabla_retrasos = df_reducido.groupby('FlightNum')['Delayed'].sum().reset_index()
tabla_retrasos = tabla_retrasos.sort_values('Delayed', ascending=False).head(5)

print(tabla_retrasos.to_string(index=False))


 FlightNum  Delayed
        15      140
        47      137
        51      137
        44      135
        16      134


- **Busca les rutes més llargues i les que acumulen més retards.**

Sin embargo aquí hemos consideraremos la distancia de los vuelos. En la primera parte, hemos agregado una condición para evitar la repetición de la misma ruta en ambas direcciones, puesto que la distancia es la misma.

In [30]:
condicion = df_reducido['Origin'] < df_reducido['Dest']
tabla_distancia = df_reducido[condicion].groupby(['Origin', 'Dest'])['Distance'].max().reset_index()
tabla_distancia = tabla_distancia.sort_values('Distance', ascending=False).head(5)

print(tabla_distancia.to_string(index=False))


Origin Dest  Distance
   EWR  HNL      4962
   ATL  HNL      4502
   HNL  ORD      4243
   KOA  ORD      4213
   HNL  MSP      3972


In [31]:
tabla_retrasos = df_reducido.groupby(['Origin', 'Dest'])['Distance'].sum().reset_index()
tabla_retrasos = tabla_retrasos.sort_values('Distance', ascending=False).head(5)

print(tabla_retrasos.to_string(index=False))


Origin Dest  Distance
   LAX  JFK    665775
   JFK  LAX    564300
   SFO  JFK    561162
   ORD  LAX    556655
   JFK  SFO    496512


- **Aporta allò que consideris rellevant.**

Llegados a este punto, sería interesante analizar qué aeropuertos presentan un mayor porcentaje de minutos de retraso acumulados en relación con el número de vuelos operados.

In [32]:
tabla_origen = df.groupby('Origin')['DepDelay'].mean().reset_index()
tabla_origen_peores = tabla_origen.sort_values(tabla_origen.columns[1], ascending=False).head(10)
print("Tabla con aeropuertos de origen con mas restrasos acumulados")
print(tabla_origen_peores.to_string(index=False))

Tabla con aeropuertos de origen con mas restrasos acumulados
Origin   DepDelay
   CMX 116.147059
   PLN  93.761905
   SPI  83.848739
   ALO  82.225806
   MQT  79.556650
   ACY  79.310345
   MOT  78.661654
   HHH  76.530055
   EGE  74.128920
   BGM  73.155340


In [33]:
tabla_origen_mejores = tabla_origen.sort_values(tabla_origen.columns[1], ascending=True).head(10)
print("Tabla con aeropuertos de origen con menos restrasos acumulados")
print(tabla_origen_mejores.to_string(index=False))

Tabla con aeropuertos de origen con menos restrasos acumulados
Origin  DepDelay
   TUP 11.000000
   YKM 22.472727
   INL 23.000000
   GST 25.075000
   WYS 25.100000
   PUB 27.000000
   IPL 28.517007
   OXR 29.268966
   CLD 29.721271
   ADK 30.509434


In [34]:
tabla_destino = df.groupby('Dest')['ArrDelay'].mean().reset_index()
tabla_destino_peores = tabla_destino.sort_values(tabla_destino.columns[1], ascending=False).head(10)
print("Tabla con aeropuertos de destino con mas restrasos acumulados")
print(tabla_destino_peores.to_string(index=False))

Tabla con aeropuertos de destino con mas restrasos acumulados
Dest  ArrDelay
 MQT 77.939252
 SPI 72.127877
 INL 66.777778
 TEX 66.266667
 ALO 65.944444
 MCN 65.674033
 CMX 63.530612
 ORD 60.028153
 EWR 59.877946
 ACK 59.167742


In [35]:
tabla_destino_mejores = tabla_destino.sort_values(tabla_destino.columns[1], ascending=True).head(10)
print("Tabla con aeropuertos de destino con menos restrasos acumulados")
print(tabla_destino_mejores.to_string(index=False))

Tabla con aeropuertos de destino con menos restrasos acumulados
Dest  ArrDelay
 TUP  6.000000
 PIR 15.000000
 BJI 20.666667
 IYK 21.112903
 BLI 22.846154
 GST 23.326087
 WYS 23.741935
 HTS 23.848485
 IPL 24.828729
 YKM 26.744681


Ahora ya sabemos que aeropuertos evitar en el futuro y cuales son mas fiables. Para terminar haré una pequeña matriz de correlación para intentar de sacarle un poco mas de jugo a este dataframe.

In [36]:
df_reducido[['DayOfWeek', 'Distance', 'AirTime', 'AvgSpeed', 'Delayed']].corr()

DayOfWeek  Distance   AirTime  AvgSpeed   Delayed
DayOfWeek   1.000000  0.007757  0.004794  0.013655 -0.002378
Distance    0.007757  1.000000  0.980491  0.655196 -0.130453
AirTime     0.004794  0.980491  1.000000  0.548011 -0.099054
AvgSpeed    0.013655  0.655196  0.548011  1.000000 -0.179221
Delayed    -0.002378 -0.130453 -0.099054 -0.179221  1.000000

Como era de esperar, la distancia y el tiempo en el aire están estrechamente relacionados. Sin embargo, la velocidad media de los vuelos no muestra una gran correlación con la duración del vuelo, lo que indica que los aviones siguen estándares de velocidad y que la distancia parece tener un peso mayor. Esto probablemente se debe a que, para vuelos más largos, se utilizan aviones más grandes con mayor potencia, lo que les permite volar un poco más rápido.

Si queremos buscar las razones de los retrasos, tendremos que analizar más detalladamente. La tabla muestra que prácticamente no existe correlación con la distancia, el tiempo en el aire ni siquiera con la velocidad media, aunque esta última es ligeramente superior, lo que indica que hay un pequeño margen en el que, si el avión puede volar más rápido, puede evitar retrasos. Es importante destacar que los valores negativos indican una relación inversa, es decir, cuanto mayor sea la velocidad media, menor será el retraso.

En cuanto al día de la semana, obviamente no muestra correlación con ninguna de las variables, ni siquiera con los retrasos, a pesar de la creencia de que los fines de semana acumulan más retrasos.

## Exercici 3
**Exporta el dataset net i amb les noves columnes a Excel.**

In [37]:
df_reducido.to_excel('Vuelos_2008.xlsx')